Project Page Enhancements

Allcharles opened the pull request:
# Enhancement of project pages

Implemented some changes to the list and detail project pages. This has enabled infinite scrolling of projects/sites, as well as filtering of the results.

## Changes

- Created `FilterComponent` to standardize creating filter inputs for pages
- Fixed broken filters/sorting on admin list pages
- Fixed spelling of taxonomic
- Added filter to project list and detail pages
- Added infinite scrolling to project list and detail pages

## Problems

- Ran problem with #451 because of slow internet

## Issues

Fixes #424 
Related #225 

## Visual Changes

### Project List Page

![project-list](https://user-images.githubusercontent.com/3955116/92598203-d4399000-f2eb-11ea-99b0-17c7dd907cd0.gif)

### Project Details Page

![project details](https://user-images.githubusercontent.com/3955116/92597692-1dd5ab00-f2eb-11ea-91f6-e5267515bd20.gif)

Allcharles committed:
Implemented infinite scrolling on projects page

Allcharles committed:
Created filter component

Implemented infinite scrolling on project details page

Allcharles committed:
Fixed broken admin pages

Allcharles committed:
Cleanup project details page implementation

Allcharles committed:
Implemented filter unit tests

Allcharles committed:
Implemented grid system for project details

Allcharles committed:
Cleanup visual bugs in map using components

Allcharles committed:
Prevent unnecessary scrolling updates

Allcharles committed:
Created project list unit tests

Allcharles committed:
Passing project details unit tests

Allcharles committed:
Added project details filter/scroll unit tests

Allcharles committed:
Fixed broken pagedTableTemplate tests

atruskie commented on src/app/components/shared/filter/filter.component.ts:
```diff
@@ -0,0 +1,37 @@
+import { debounceTime, distinctUntilChanged } from "rxjs/operators";
+
+@Component({
+  selector: "baw-filter",
```
might do with a more descriptive name?

`baw-text-filter`
`baw-search-filter`
`baw-page-search`

Filter is somewhat ambiguous because we have API filters, and filters on other (non-text) fields.

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -55,10 +64,56 @@ class DetailsComponent extends PageComponent implements OnInit {
-    );
+
+    this.loading = true;
+    this.sites$
```
is the dollar suffix a convention?

Allcharles replied:
There is no set convention, however its a simple (library specific method) of knowing what type of variable `sites$` is. The $ means its an observable: https://stackoverflow.com/questions/43083577/what-does-sign-at-the-end-of-function-name-indicate.

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -55,10 +64,56 @@ class DetailsComponent extends PageComponent implements OnInit {
+  public onScroll() {
+    this.page++;
+    this.loading = true;
+    this.sites$.next();
```
nice

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -33,19 +37,24 @@ export const projectMenuItemActions = [
+  private filter: InnerFilter<ISite>;
 
-  constructor(private route: ActivatedRoute) {
+  constructor(private route: ActivatedRoute, private api: SitesService) {
```
`api` is an inaccurate name for the SiteService. Try: `sitesService`

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -55,10 +64,56 @@ class DetailsComponent extends PageComponent implements OnInit {
+            sites.length === 0 ||
+            sites[0].getMetadata().paging.maxPage === this.page;
+        })
+      );
```
I think a lot of this logic is quite general and could be abstracted.

The site service should probably offer a method like `pagedFilter` that return a subject.

The `next` method can accept parameters, so it could accept `page` and `filter` (and the page reset on filter change logic can be neatly packaged within).

Further, such functionality could be made as a reusable mixin that any of our API<Filter> services could include.

atruskie commented on src/app/components/projects/pages/list/list.component.ts:
```diff
@@ -1,64 +1,122 @@
+            projects.length === 0 ||
+            projects[0].getMetadata().paging.maxPage === this.page;
+        })
+      );
```
same comments as previous

atruskie commented on src/app/components/shared/filter/filter.component.ts:
```diff
@@ -0,0 +1,37 @@
+
+  public ngOnInit() {
+    this.input$
+      .pipe(debounceTime(500), distinctUntilChanged())
```
might be useful to have kind of common global input debounce time? `BAW_KEYBOARD_DEBOUNCE`

atruskie commented on src/app/helpers/embedGoogleMaps/embedGoogleMaps.ts:
```diff
@@ -27,9 +27,9 @@ export async function embedGoogleMaps(key?: string) {
+      } else if (count > 10) {
         console.error("Failed to load google maps.");
-        reject();
+        reject("Google Maps API Bundle took too long to download.");
```
"try refreshing the page"?

Allcharles replied:
Official error message already does that, that error message will be printed to the console for developers.

atruskie reviewed (changes requested):
The filter component is really nice!
On the whole really great work. Could just use a little tightening up in the component/controller code for the paging.

Allcharles committed:
Rename FilterComponent

Allcharles committed:
Merge branch 'master' into infinite-scrolling

Allcharles committed:
Fix review comments

Allcharles committed:
QSP Filtering (#481)

* Extracted duplicate logic to scrollTemplate

Fixed broken styling for mobile devices in project details page

* Saved filter to url

* Added page tracking

Removed infinite scrolling

* Fixed QSP bug

QSP values were not being reset because of a runaway rxjs observable. Linter should catch this issue, however its failing to detect rxjs problems

* Set pagination defaults

NgbPagination exposes a config service which can be used to set defaults for the component. This reduces duplicate code in the paginationTemplate extending components.

* Initial solution issue #477

Seems to have some performance issues

* Map component markers now immutable

* Removed duplicate logic in paginationTemplate

* Extract map marker logic to separate component

* Added site-map tests

* Implemented paginationTemplate unit tests

Allcharles committed:
Merge branch 'master' into infinite-scrolling
